File: app/src/main/java/org/supla/android/lib/SuplaChannelState.java
package org.supla.android.lib;

/*
 Copyright (C) AC SOFTWARE SP. Z O.O.

 This program is free software; you can redistribute it and/or
 modify it under the terms of the GNU General Public License
 as published by the Free Software Foundation; either version 2
 of the License, or (at your option) any later version.

 This program is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 GNU General Public License for more details.

 You should have received a copy of the GNU General Public License
 along with this program; if not, write to the Free Software
 Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
 */

import android.annotation.SuppressLint;
import android.content.Context;
import android.content.res.Resources;

import org.supla.android.R;

import java.io.Serializable;

@SuppressWarnings("unused")
public class SuplaChannelState implements Serializable {

    public static final int FIELD_IPV4 = 0x0001;
    public static final int FIELD_MAC = 0x0002;
    public static final int FIELD_BATTERYLEVEL = 0x0004;
    public static final int FIELD_BATTERYPOWERED = 0x0008;
    public static final int FIELD_WIFIRSSI = 0x0010;
    public static final int FIELD_WIFISIGNALSTRENGTH = 0x0020;
    public static final int FIELD_BRIDGENODESIGNALSTRENGTH = 0x0040;
    public static final int FIELD_UPTIME = 0x0080;
    public static final int FIELD_CONNECTIONUPTIME = 0x0100;
    public static final int FIELD_BATTERYHEALTH = 0x0200;
    public static final int FIELD_BRIDGENODEONLINE = 0x0400;
    public static final int FIELD_LASTCONNECTIONRESETCAUSE = 0x0800;
    public static final int FIELD_LIGHTSOURCELIFESPAN = 0x1000;
    public static final int FIELD_LIGHTSOURCELIFEOPERATINGTIME = 0x2000;

    private int ChannelID;
    private int fields;
    private int defaultIconField;
    private Integer ipv4;
    private byte[] macAddress;
    private Byte batteryLevel;
    private Boolean batteryPowered;
    private Byte wiFiRSSI;
    private Byte wiFiSignalStrength;
    private Boolean bridgeNodeOnline;
    private Byte bridgeNodeSignalStrength;
    private Long uptime;
    private Long connectionUptime;
    private Byte batteryHealth;
    private Byte lastConnectionResetCause;
    private Integer lightSourceLifespan;
    private Float lightSourceLifespanLeft;
    private Integer lightSourceOperatingTime;

    public SuplaChannelState(int ChannelID, int fields, int defaultIconField,
                             int ipv4, byte[] macAddress, byte batteryLevel,
                             byte batteryPowered, byte wiFiRSSI, byte wiFiSignalStrength,
                             byte bridgeNodeOnline, byte bridgeNodeSignalStrength, int uptime,
                             int connectionUptime, byte batteryHealth,
                             byte lastConnectionResetCause, int lightSourceLifespan,
                             int lightSourceLifespanLeft) {

        this.ChannelID = ChannelID;
        this.fields = fields;
        this.defaultIconField = defaultIconField;

        if ((fields & FIELD_IPV4) > 0) {
            this.ipv4 = ipv4;
        }

        if ((fields & FIELD_MAC) > 0) {
            this.macAddress = macAddress;
        }

        if ((fields & FIELD_BATTERYLEVEL) > 0
                && batteryLevel >= 0
                && batteryLevel <= 100) {
            this.batteryLevel = batteryLevel;
        }

        if ((fields & FIELD_BATTERYPOWERED) > 0) {
            this.batteryPowered = batteryPowered > 0;
        }

        if ((fields & FIELD_WIFIRSSI) > 0) {
            this.wiFiRSSI = wiFiRSSI;
        }

        if ((fields & FIELD_WIFISIGNALSTRENGTH) > 0
                && wiFiSignalStrength >= 0
                && wiFiSignalStrength <= 100) {
            this.wiFiSignalStrength = wiFiSignalStrength;
        }

        if ((fields & FIELD_BRIDGENODEONLINE) > 0) {
            this.bridgeNodeOnline = bridgeNodeOnline > 0;
        }

        if ((fields & FIELD_BRIDGENODESIGNALSTRENGTH) > 0
                && bridgeNodeSignalStrength >= 0
                && bridgeNodeSignalStrength <= 100) {
            this.bridgeNodeSignalStrength = bridgeNodeSignalStrength;
        }

        if ((fields & FIELD_UPTIME) > 0) {
            this.uptime = uptime & 0x00000000ffffffffL;
        }

        if ((fields & FIELD_CONNECTIONUPTIME) > 0) {
            this.connectionUptime = connectionUptime & 0x00000000ffffffffL;
        }

        if ((fields & FIELD_BATTERYHEALTH) > 0
                && batteryHealth >= 0
                && batteryHealth <= 100) {
            this.batteryHealth = batteryHealth;
        }

        if ((fields & FIELD_LASTCONNECTIONRESETCAUSE) > 0) {
            this.lastConnectionResetCause = lastConnectionResetCause;
        }

        if ((fields & FIELD_LIGHTSOURCELIFESPAN) > 0) {
            this.lightSourceLifespanLeft = lightSourceLifespanLeft / 100f;
            this.lightSourceLifespan = lightSourceLifespan;
        }

        if ((fields & FIELD_LIGHTSOURCELIFEOPERATINGTIME) > 0) {
            this.lightSourceLifespanLeft = null;
            this.lightSourceOperatingTime = lightSourceLifespanLeft;
        }
    }

    public int getChannelID() {
        return ChannelID;
    }

    public Integer getIpv4() {
        return ipv4;
    }

    @SuppressLint("DefaultLocale")
    public String getIpv4String() {
        int ipv4 = this.ipv4.intValue();
        return String.format(
                "%d.%d.%d.%d",
                (ipv4 & 0xff),
                (ipv4 >> 8 & 0xff),
                (ipv4 >> 16 & 0xff),
                (ipv4 >> 24 & 0xff));
    }

    public byte[] getMacAddress() {
        return macAddress;
    }

    public String getMacAddressString() {
        StringBuilder sb = new StringBuilder(macAddress.length * 3 - 1);
        for (byte b : macAddress) {
            if (sb.length() > 0) {
                sb.append(":");
            }
            sb.append(String.format("%02x", b));
        }

        return sb.toString();
    }

    public Byte getBatteryLevel() {
        return batteryLevel;
    }

    public String getBatteryLevelString() {
        return String.format("%d%%", getBatteryLevel());
    }

    public Boolean isBatteryPowered() {
        return batteryPowered;
    }

    public String getBatteryPoweredString(Context context) {
        Resources r = context.getResources();
        return r.getString(isBatteryPowered() ? R.string.yes : R.string.no);
    }

    public Byte getWiFiSignalStrength() {
        return wiFiSignalStrength;
    }

    public String getWiFiSignalStrengthString() {
        return String.format("%d%%", getWiFiSignalStrength());
    }

    public Byte getWiFiRSSI() {
        return wiFiRSSI;
    }

    public String getWiFiRSSIString() {
        return String.format("%d", getWiFiRSSI());
    }

    public Byte getBridgeNodeSignalStrength() {
        return bridgeNodeSignalStrength;
    }

    public String getBridgeNodeSignalStrengthString(Context context) {
        return String.format("%d%%", getBridgeNodeSignalStrength());
    }

    private String getUptimeString(Long uptime, Context context) {
        long secs = uptime == null ? 0 : uptime;

        return String.format("%d %s %02d:%02d:%02d",
                uptime / 86400,
                context.getResources().getString(R.string.days).toLowerCase(),
                uptime % 86400 / 3600,
                uptime % 86400 % 3600 / 60,
                uptime % 86400 % 3600 % 60);
    }

    public Long getUptime() {
        return uptime;
    }

    public String getUptimeString(Context context) {
        return getUptimeString(uptime, context);
    }

    public Long getConnectionUptime() {
        return connectionUptime;
    }

    public String getConnectionUptimeString(Context context) {
        return getUptimeString(connectionUptime, context);
    }

    public int getFields() {
        return fields;
    }

    public int getDefaultIconField() {
        return defaultIconField;
    }

    public Boolean isBridgeNodeOnline() {
        return bridgeNodeOnline;
    }

    public String getBridgeNodeOnlineString(Context context) {
        Resources r = context.getResources();
        return r.getString(isBridgeNodeOnline() ? R.string.yes : R.string.no);
    }

    public Byte getBatteryHealth() {
        return batteryHealth;
    }

    public String getBatteryHealthString() {
        return String.format("%d%%", getBatteryHealth());
    }

    public Byte getLastConnectionResetCause() {
        return lastConnectionResetCause;
    }

    public String getLastConnectionResetCauseString() {
        switch (getLastConnectionResetCause()) {
            default:
                return String.format("%d", getLastConnectionResetCause());
        }
    }

    public Integer getLightSourceLifespan() {
        return lightSourceLifespan;
    }

    public Float getLightSourceLifespanLeft() {
        return lightSourceLifespanLeft;
    }

    public String getLightSourceLifespanString() {

        Float percentLeft = null;

        if (getLightSourceLifespanLeft() != null) {
            percentLeft = getLightSourceLifespanLeft();
        } else if (getLightSourceOperatingTimePercentLeft() != null) {
            percentLeft = getLightSourceOperatingTimePercentLeft();
        }

        if (percentLeft != null) {
            return String.format("%dh (%.2f%%)",
                    getLightSourceLifespan(),
                    percentLeft);
        }

        return String.format("%dh", getLightSourceLifespan());
    }

    public Integer getLightSourceOperatingTime() {
        return lightSourceOperatingTime;
    }

    public Float getLightSourceOperatingTimePercent() {
        if (getLightSourceLifespan() != null && getLightSourceLifespan().intValue() > 0) {
            return getLightSourceOperatingTime() / 36f / getLightSourceLifespan();
        }
        return null;
    }

    public Float getLightSourceOperatingTimePercentLeft() {
        Float percent = getLightSourceOperatingTimePercent();
        if (percent != null) {
            return 100 - percent;
        }
        return null;
    }

    public String getLightSourceOperatingTimeString() {
        long timeSec = getLightSourceOperatingTime().longValue();

        return String.format("%02dh %02d:%02d",
                timeSec / 3600,
                timeSec % 3600 / 60,
                timeSec % 3600 % 60);

    }
}

File: app/src/main/java/org/supla/android/lib/SuplaChannelState.kt
package org.supla.android.lib
/*
 Copyright (C) AC SOFTWARE SP. Z O.O.

 This program is free software; you can redistribute it and/or
 modify it under the terms of the GNU General Public License
 as published by the Free Software Foundation; either version 2
 of the License, or (at your option) any later version.

 This program is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 GNU General Public License for more details.

 You should have received a copy of the GNU General Public License
 along with this program; if not, write to the Free Software
 Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
 */

import org.supla.android.R
import org.supla.android.data.source.local.entity.ChannelStateEntity
import org.supla.android.extensions.ipV4String
import org.supla.android.tools.UsedFromNativeCode
import org.supla.core.shared.extensions.ifTrue
import org.supla.core.shared.extensions.toHex
import org.supla.core.shared.infrastructure.LocalizedString
import java.io.Serializable
import java.util.Locale

@UsedFromNativeCode
class SuplaChannelState(
  val channelId: Int,
  private val fields: Int,
  private val rawDefaultIconField: Int,
  private val rawIpv4: Int,
  private val rawMacAddress: ByteArray,
  private val rawBatteryLevel: Byte,
  private val rawBatteryPowered: Byte,
  private val rawWifiRssi: Byte,
  private val rawWifiSignalStrength: Byte,
  private val rawBridgeNodeOnline: Byte,
  private val rawBridgeNodeSignalStrength: Byte,
  private val rawUptime: Int,
  private val rawConnectionUptime: Int,
  private val rawBatteryHealth: Byte,
  private val rawLastConnectionResetCause: Byte,
  private val rawLightSourceLifespan: Int,
  private val rawLightSourceLifespanLeft: Int
) : Serializable {

  val ipV4: String?
    get() = hasField(FIELD_IPV4).ifTrue { rawIpv4.ipV4String }

  val macAddress: String?
    get() = hasField(FIELD_MAC).ifTrue { rawMacAddress.toHex(":") }

  val batteryLevel: Int?
    get() = hasField(FIELD_BATTERYLEVEL).ifTrue { rawBatteryLevel.toInt() }

  val batteryLevelString: String?
    get() = batteryLevel?.let { "$it%" }

  val batteryPowered: Boolean?
    get() = hasField(FIELD_BATTERYPOWERED).ifTrue { rawBatteryPowered > 0 }

  val batterPoweredString: LocalizedString?
    get() = batteryPowered?.let { LocalizedString.WithResource(if (it) R.string.yes else R.string.no) }

  val wifiRssi: Int?
    get() = hasField(FIELD_WIFIRSSI).ifTrue { rawWifiRssi.toInt() }

  val wifiRssiString: String?
    get() = wifiRssi?.let { "$it" }

  val wifiSignalStrength: Int?
    get() = hasField(FIELD_WIFISIGNALSTRENGTH).ifTrue { rawWifiSignalStrength.toInt() }

  val wifiSignalStrengthString: String?
    get() = wifiSignalStrength?.let { "$it%" }

  val bridgeNodeOnline: Boolean?
    get() = hasField(FIELD_BRIDGENODEONLINE).ifTrue { rawBridgeNodeOnline > 0 }

  val bridgeNodeSignalStrength: Int?
    get() = hasField(FIELD_BRIDGENODESIGNALSTRENGTH).ifTrue { rawBridgeNodeSignalStrength.toInt() }

  val bridgeNodeSignalStrengthString: String?
    get() = bridgeNodeSignalStrength?.let { "$it%" }

  val uptime: Int?
    get() = hasField(FIELD_UPTIME).ifTrue { rawUptime }

  val uptimeString: LocalizedString?
    get() = uptime?.let { getUptimeString(it) }

  val connectionUptime: Int?
    get() = hasField(FIELD_CONNECTIONUPTIME).ifTrue { rawConnectionUptime }

  val connectionUptimeString: LocalizedString?
    get() = connectionUptime?.let { getUptimeString(it) }

  val batteryHealth: Int?
    get() = hasField(FIELD_BATTERYHEALTH).ifTrue { rawBatteryHealth.toInt() }

  val batteryHealthString: String?
    get() = batteryHealth?.let { "$it%" }

  val lastConnectionResetCause: Int?
    get() = hasField(FIELD_LASTCONNECTIONRESETCAUSE).ifTrue { rawLastConnectionResetCause.toInt() }

  val lastConnectionResetCauseString: LocalizedString?
    get() = lastConnectionResetCause?.let { cause ->
      val causeResources = listOf(
        R.string.lastconnectionresetcause_unknown,
        R.string.lastconnectionresetcause_activity_timeout,
        R.string.lastconnectionresetcause_wifi_connection_lost,
        R.string.lastconnectionresetcause_server_connection_lost
      )
      causeResources.getOrNull(cause)?.let { LocalizedString.WithResource(it) } ?: LocalizedString.Constant("$cause")
    }

  val lightSourceLifespan: Int?
    get() = hasField(FIELD_LIGHTSOURCELIFESPAN).ifTrue { rawLightSourceLifespan }

  val lightSourceLifespanString: String?
    get() = lightSourceLifespan?.let { lightSourceLifespan ->
      val left = lightSourceLifespanLeft ?: lightSourceOperatingTimePercentLeft

      if (left != null) {
        String.format(Locale.getDefault(), "%dh (%.2f%%)", lightSourceLifespan, left)
      } else {
        String.format(Locale.getDefault(), "%dh", lightSourceLifespan)
      }
    }

  val lightSourceLifespanLeft: Float?
    get() = if (hasField(FIELD_LIGHTSOURCELIFESPAN) && !hasField(FIELD_LIGHTSOURCELIFEOPERATINGTIME)) {
      rawLightSourceLifespanLeft / 100f
    } else {
      null
    }

  val lightSourceOperatingTime: Int?
    get() = hasField(FIELD_LIGHTSOURCELIFEOPERATINGTIME).ifTrue { rawLightSourceLifespanLeft }

  val lightSourceOperatingTimeString: String?
    get() = lightSourceOperatingTime?.let {
      String.format(Locale.getDefault(), "%02dh %02d:%02d", it / 3600, it % 3600 / 60, it % 60)
    }

  val lightSourceOperatingTimePercent: Float? =
    lightSourceOperatingTime?.let { lightSourceOperatingTime ->
      lightSourceLifespan?.let { lightSourceLifespan ->
        (lightSourceLifespan > 0).ifTrue { lightSourceOperatingTime / 36f / lightSourceLifespan }
      }
    }

  val lightSourceOperatingTimePercentLeft: Float? =
    lightSourceOperatingTimePercent?.let { 100 - it }

  fun toEntity(profileId: Long): ChannelStateEntity {
    return ChannelStateEntity(
      batteryHealth = batteryHealth,
      batteryLevel = batteryLevel,
      batteryPowered = batteryPowered,
      bridgeNodeOnline = bridgeNodeOnline,
      bridgeNodeSignalStrength = bridgeNodeSignalStrength,
      connectionUptime = connectionUptime,
      ipV4 = ipV4,
      lastConnectionResetCause = lastConnectionResetCause,
      lightSourceLifespan = lightSourceLifespan,
      lightSourceLifespanLeft = lightSourceLifespanLeft,
      lightSourceOperatingTime = lightSourceOperatingTime,
      macAddress = macAddress,
      uptime = uptime,
      wifiRssi = wifiRssi,
      wifiSignalStrength = wifiSignalStrength,
      channelId = channelId,
      profileId = profileId
    )
  }

  private fun hasField(field: Int): Boolean = (fields and field) > 0

  private fun getUptimeString(time: Int): LocalizedString = LocalizedString.WithResourceIntIntIntInt(
    id = R.string.channel_state_uptime,
    arg1 = time / 86400,
    arg2 = time % 86400 / 3600,
    arg3 = time % 3600 / 60,
    arg4 = time % 60
  )

  companion object {
    const val FIELD_IPV4: Int = 0x0001
    const val FIELD_MAC: Int = 0x0002
    const val FIELD_BATTERYLEVEL: Int = 0x0004
    const val FIELD_BATTERYPOWERED: Int = 0x0008
    const val FIELD_WIFIRSSI: Int = 0x0010
    const val FIELD_WIFISIGNALSTRENGTH: Int = 0x0020
    const val FIELD_BRIDGENODESIGNALSTRENGTH: Int = 0x0040
    const val FIELD_UPTIME: Int = 0x0080
    const val FIELD_CONNECTIONUPTIME: Int = 0x0100
    const val FIELD_BATTERYHEALTH: Int = 0x0200
    const val FIELD_BRIDGENODEONLINE: Int = 0x0400
    const val FIELD_LASTCONNECTIONRESETCAUSE: Int = 0x0800
    const val FIELD_LIGHTSOURCELIFESPAN: Int = 0x1000
    const val FIELD_LIGHTSOURCELIFEOPERATINGTIME: Int = 0x2000
  }
}
